Add getallpvrthumb to webservice

Ignoble61 opened the pull request:
I was envious of how you add skin helper's cached artwork directly to your PVR recordings widget, and wanted to recreate it in smart(ish) widgets which I'm slowly updating...

This adds 'getallpvrthumb' to the web service, which returns the dictionary of artwork that skin helper has scraped, rather than an individual image (so that smart(ish) can then add the images directly - https://github.com/Ignoble61/service.smartish.widgets/commit/410471664e4ef9a9492b14901ce7442d60f0321d)

You may not like how I'm determining whether to return a single image, or whether to return the dictionary - if there's anyway you'd prefer it to work let me know and I'll re-write it with pleasure. Also, within the getPVRThumbs function I've unquoted the channel and title, as skin helper was searching for the quoted version (e.g. with spaces replaced with %20's, and so on). Doesn't seem to matter how I encode/quote these when sending from smart(ish), it appears to need unquoting within helper.

marcelveldt commented:
Oops, I have changed some stuff around the other day that is now conflicting with your changes. I can't auto merge this. Do you mind take a look at the upstream changes so we can get this merged ?

Nice addition btw, use the webservice to exchange info between skinhelper and other addons :-)

marcelveldt commented:
Take a look at the getCompareTitle method I created... I saw some redundant ocurring code to strip the chars to perform a compare and I moved that into a separate method. Most likely that is the thing that is now conflicting.

Ignoble61 commented:
I'll rebase this evening. To be clear, you're happy enough with how I'm testing whether to return the image itself or the dictionary...? (I'm not 100% happy with it just because it complicates some nice clear code, so if you have a better idea please say now so I can rewrite whilst I rebase!)

marcelveldt commented:
Well, the only other way I see is to add a param to "send_headers" that the request is head only and just call the "send_headers" method and do the logic if the image itself is sent or just evaluated in the method itself instead of in the head and get methods...

That will also be a bit more flexible in the future if the webservice is to be extended with more methods.

something like this:

def send_headers(self, sendContent=False):
    #some stuff here and send the image or just the headers based on the optional param

Ignoble61 commented:
I'm not really up on the SimpleHttpServer (I looking into it for smart(ish) but decided to use sockets instead) so I could be completely wrong here, but the method I added doesn't just send headers - it sends the dict (which it has placed in the image parameter) in the do_GET() function which I believe is the 'body' of the response. I wouldn't have thought it was possible to send the dict in the headers, but tell me if I'm wrong and I'll do some reading this evening on the relevant methods :)

marcelveldt commented:
You should be fine adding it only to the get portion of the webservice. The only reason I've also added the HEAD method was for Kodi's texture cache which appears to be Always performing a HEAD request to the url to determine if the content had changed before it downloads an image.

So, the only thing the HEAD request will return is the headers with the content type, last modified etc.
The GET will return the actual content (body), so in case of the images stuff, it returns the image. In your situation it will return the dict (or maybe even json ?)

Ignoble61 commented:
OK, I've rebased (hopefully correctly - I need to double-check it's all working fine, though smart(ish) is communicating correctly).

As you can see, I'm sending the dict as plaintext - firstly, so I don't have to start converting it to other formats (though json would make sense, and some processing is needed as HttpBasicServer doesn't play nicely with sending unicode), and second so that I can get the original dict back very easily on the other end with urlparse.parse_qs.

send_headers() seems a good place to do this - first of all its where you grab the data for the current getpvrthumbs call (and the new method still needs to grab that data), and second it sets the correct headers so that plaintext is expected (requests could probably handle this not being there, but better to be correct, and it definitely needs the 200 response).

What I thought you might take issue with is that I've added a second return parameter - 'multi' - to send_headers(). If it's set, it's returning multiple images, not just a single image, and so gets treated differently in do_GET()

marcelveldt commented:
code looks fine... One question.. does the urllib.unquote play nice with unicode data ?
All strings are converted to unicode by the script

Ignoble61 commented:
I tested it with the following:-

testString1 = urllib.unquote(u"œ∑´®†ƒ©˙∆˚˙ƒ©∆˚∫~√√ßsßåßà")
testString2 = urllib.unquote(u"œ∑´®†ƒ©˙∆˚˙ƒ©∆˚∫~√√ßsßåßà".encode("utf-8"))
testString3 = urllib.unquote("œ∑´®†ƒ©˙∆˚˙ƒ©∆˚∫~√√ßsßåßà")
testString4 = urllib.unquote("œ∑´®†ƒ©˙∆˚˙ƒ©∆˚∫~√√ßsßåßà".decode("utf-8"))

Which I admit isn't a great test, but it didn't throw any errors...
